### PR TITLE
Add "buf" and "win" to nvim_get_option_value and use them in vim.bo and vim.wo

### PR DESCRIPTION
**Pull request by gpanders**
Add "buf" and "win" options to `nvim_get_option_value`. Use `nvim_get_option_value` and `nvim_set_option_value` for `vim.bo` and `vim.wo`.

`nvim_get_option_value` and `nvim_set_option_value` better handle unsetting local options. For instance, this is currently not possible:

    vim.bo.tagfunc = nil

This does not work because 'tagfunc' is marked as "local to buffer" and does not have a fallback global option. However, using :setlocal *does* work as expected

    :setlocal tagfunc=

`nvim_set_option_value` behaves more like :set and :setlocal (by design), so using these as the underlying API functions beneath vim.bo and vim.wo makes those two tables act more like :setlocal. Note that vim.o *already* uses `nvim_set_option_value` under the hood, so that vim.o behaves like :set. Thus with this change the `vim.o`, `vim.go`, `vim.bo`, and `vim.wo` family of tables should perfectly match the usage as expected by `:set`, `:setglobal`, and `:setlocal`.

### CODE REVIEW

File: src/nvim/api/private/helpers.c [hunk 1]
```diff
@@ -1039,6 +1039,45 @@ Object copy_object(Object obj)
   }
 }
 
+int get_option_value_for(char *key, long *numval, char **stringval, int opt_flags, int opt_type,
```
Comment (lewis6991) [inline src/nvim/api/private/helpers.c:1042]: This is *very* similar to `set_option_value_for`. Is there any reasonable way we can unify them?

File: src/nvim/api/vim.c [hunk 2]
```diff
@@ -737,9 +738,44 @@ Object nvim_get_option_value(String name, Dict(option) *opts, Error *err)
     goto end;
   }
 
+  int opt_type = SREQ_GLOBAL;
+  void *from = NULL;
+
+  if (opts->win.type == kObjectTypeInteger) {
```
Comment (lewis6991) [inline src/nvim/api/vim.c:744]: Again, the argument validation is quite similar to `nvim_set_option_value`. Can we unify?

File: test/functional/api/vim_spec.lua [hunk 3]
```diff
@@ -1457,6 +1457,17 @@ describe('API', function()
...
+
+    it('get buffer or window-local options', function()
+      nvim('command', 'new')
+      local buf = nvim('get_current_buf').id
+      nvim('buf_set_option', buf, 'tagfunc', 'foobar')
+      eq('foobar', nvim('get_option_value', 'tagfunc', {buf = buf}))
+
+      local win = nvim('get_current_win').id
+      nvim('win_set_option', win, 'number', true)
+      eq(true, nvim('get_option_value', 'number', {win = win}))
```
Comment (lewis6991) [inline test/functional/api/vim_spec.lua:1469]: I think it is worth adding some coverage for equiv of `setlocal <winopt>` (`{win=win, scope='local'}`) as that is the main difference to `nvim_win_get_option` and one of the main motivations for adding this change.

**Review by lewis6991 — COMMENTED** [review]
LGTM other than some potential refactoring and test coverage.

File: src/nvim/api/private/helpers.c (hunk 1 above)
Comment (gpanders) [reply src/nvim/api/private/helpers.c:1042]: Imo the effort to design a common abstraction for just these two functions is not worth it.

Comment (lewis6991) [reply src/nvim/api/private/helpers.c:1042]: Design? Come-on, it isn't that difficult. It's just basic refactoring.

**Comment by gpanders** [comment]
Separate commits are intentional, do not squash when merging.

**Review by justinmk — APPROVED** [review]
Hugely appreciated

**Comment by shadmansaleh** [comment]
There's a strange bug occurring after this https://github.com/nvim-lualine/lualine.nvim/issues/741 . Any idea why ?

To reproduce you can use this as minimal init

```lua
function stl() return vim.bo.filetype end
vim.go.stl = '%!v:lua.stl()'
```

**Comment by clason** [comment]
This commit has also introduced a regression in telescope (around insert mode and cursor position when closing a picker), which is similarly baffling.

**Comment by rapan931** [comment]
OK! I created issue to neovim https://github.com/neovim/neovim/issues/19381

**Pull request merged**